### PR TITLE
v6 — Improve redux connector performances

### PR DESCRIPTION
**Pull request by ooflorent**
Use Redux's memoization API to improve `ConnectedForm` performances by avoiding closure re-creations.

**NOTE:** It introduces 2 breaking changes:

- `react-redux` MUST match `>=4.3` (previously `>=4.0`)
- `form`, `touchOnBlur` and `touchOnChange` can't change over time. In order to prevent WTF behaviors, we could add a warning when `NODE_ENV` is not `production` if these props change.

**Disclaimer:** I haven't tested the examples since they rely on the latest published version on Redux Form instead of the local one.

### CODE REVIEW

**Comment by ooflorent** [comment]
Fixed linter errors

File: src/Field.js [hunk 1]
```diff
@@ -54,8 +54,10 @@ const createField = ({ deepEqual, getIn }) => {
     }
 
     render() {
-      const { ConnectedField } = this
-      return <ConnectedField {...this.props} ref="connected"/>
+      return createElement(this.ConnectedField, {
+        ...this.props,
+        ref: 'connected'
+      })
```
Comment (erikras) [inline src/Field.js]: Anything in particular gained by avoiding JSX here? or just using the `createElement()` since we already have it in scope?

**Comment by erikras** [comment]
> we could add a warning when NODE_ENV is not production if these props change.

`invariant` is already a dep. Use it for this.

File: src/Field.js (hunk 1 above)
Comment (ooflorent) [inline src/Field.js]: I replaced JSX in favor of `createElement` to take control over what transpilation is doing. The end result is cleaner in my opinion because by reading the code we immediately understand what react is doing. I've done these changes in a separate commit so it could be easily reverted.

Note: It is also how `react-redux` instantiate React elements.

Comment (erikras) [inline src/Field.js]: 👍 Good enough. Was just curious.

**Comment by erikras** [comment]
The benefit of this optimization just clicked in my mind, and now I understand it. That's a _lot_ of binding we will be avoiding on every render!  🎉 🎉  Definitely worth the cost of prohibiting the mutation of those three props!

**Comment by ooflorent** [comment]
I've updated the PR to avoid creating the action map on each render.

**Comment by erikras** [comment]
> Disclaimer: I haven't tested the examples since they rely on the latest published version on Redux Form instead of the local one.

You should _probably_ do that before we merge this, as it's a pretty big change. It's just a matter of copying the `lib` (and maybe `es`) folders over the ones in `node_modules/redux-form` in every example.

However, that the tests are all passing is promising. That's the nice thing about having so many tests.

---

Actually, it never occurred to me that now that the examples are in the same branch as the library code, that they could just `import { reduxForm } from '../../reduxForm'`. I think that's how most libs do their examples. I wonder if that would be better than the current set up. I sort of like how I can confidently say that "this example project is using _this_ published version". Your thoughts?

**Comment by ooflorent** [comment]
I see several options:

- using `npm link` (which causes issues when deps got duplicated)
- using relative paths
- using webpack aliases with relative paths

> I sort of like how I can confidently say that "this example project is using this published version"

Since each release has a git tag it is not an issue.

**Comment by erikras** [comment]
Published as [`v6.0.0-alpha.13`](https://github.com/erikras/redux-form/releases/tag/v6.0.0-alpha.13).

**Comment by erikras** [comment]
> * using `npm link` (which causes issues when deps got duplicated)

Yes, `npm link` sucks. 👎 

> * using relative paths
> * using webpack aliases with relative paths

The aliases don't excite me that much, but the problem with relative paths is that the examples don't get to inherit any of `redux-form`'s deps. i.e. they would have to include them all. Still might be worth it, though.
